File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

import math


class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        def dfs(acc: int, node: Optional[TreeNode]) -> tuple[int, int]:
            if node is None:
                return (acc, 0)

            (acc, left) = dfs(acc, node.left)
            (acc, right) = dfs(acc, node.right)

            acc = max(acc, left + right + node.val)

            value = max(left, right) + node.val
            if value < 0:
                return (acc, 0)
            else:
                return (acc, value)

        return dfs(-1000, root)[0]

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
  TreeNode *lowestCommonAncestor(TreeNode *root, TreeNode *p, TreeNode *q) {
    auto l = p->val <= q->val ? p : q;
    auto r = p->val >= q->val ? p : q;
    if (l->val <= root->val && root->val <= r->val)
      return root;
    if (l->val <= root->val && r->val <= root->val)
      return lowestCommonAncestor(root->left, l, r);
    if (l->val >= root->val && r->val >= root->val)
      return lowestCommonAncestor(root->right, l, r);
    throw std::runtime_error("p and q are not in the same subtree");
  }
};

